PasswordObfuscator should not assume PASSWORD is the last item in the…

bereng opened the pull request:
… WITH clause

bereng committed:
PasswordObfuscator should not assume PASSWORD is the last item in the WITH clause

bereng commented:
CI [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/490/workflows/1e6ee30c-b6eb-418c-b26f-e32188e04ba0)
CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/490/workflows/6db5e225-9ec8-415e-8f6a-3fd64d19fafd)

CI lgtm given auth test is currently broken

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -35,12 +35,14 @@
 import org.apache.cassandra.exceptions.ConfigurationException;
 import org.apache.cassandra.exceptions.PreparedQueryNotFoundException;
+import org.apache.cassandra.exceptions.SyntaxException;
 import org.apache.cassandra.exceptions.UnauthorizedException;
```
There is one unused import statement` import java.util.Map;`

bereng replied:
Right!

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/cql3/PasswordObfuscator.java:
```diff
@@ -35,6 +40,18 @@ public String obfuscate(String sourceString)
+        if (!pass.isPresent() || pass.get().isEmpty())
+            return query;
+        else
+            return query.replaceAll(pass.get(), PasswordObfuscator.OBFUSCATION_TOKEN);
```
Please add java doc to the two methods

ekaterinadimitrova2 replied:
Also, we will have to update NEWS.txt and the docs because now only the password will be obfuscated

bereng replied:
Good call! I guess you meant CHANGES.txt. Let me know otherwise.

ekaterinadimitrova2 replied:
Also NEWS.txt because there will be a visible change for the users. The entries will obfuscate only the password and the DCL statements won't be that useless anymore

bereng replied:
Oh right! bc this was released on 4.0. Right.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/cql3/QueryEvents.java:
```diff
@@ -81,9 +73,9 @@ public void notifyQuerySuccess(CQLStatement statement,
         try
         {
-            final String possiblyObfuscatedQuery = listeners.size() > 0 ? possiblyObfuscateQuery(statement, query) : query;
+            final String maybeObfuscatedQuery = listeners.size() > 0 ? maybeObfuscatePassword(statement, query) : query;
```
To me personally this name change doesn't bring any value and just adds noise

bereng replied:
If you search for methods named `maybe*` in the codebase there are 600+. For `possibly*` there a 5 in external jars. So I see it like a non-written rule to name them as `maybe*` to align to rest of the project's style.

ekaterinadimitrova2 replied:
Fair point

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/cql3/QueryEvents.java:
```diff
@@ -250,17 +242,27 @@ public void notifyPrepareFailure(@Nullable CQLStatement statement, String query,
         return !listeners.isEmpty();
     }
 
+    @SuppressWarnings("unused")
```
I would prefer for us to concentrate on the problem and refrain from adding an unrelated suppression

bereng replied:
Personal taste imo. I've had other committer tell me the opposite. So now that it's there lets leave it?

ekaterinadimitrova2 replied:
I would prefer to remove it tbh

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/audit/AuditLoggerAuthTest.java:
```diff
@@ -295,8 +330,11 @@ private static void assertSource(AuditLogEntry logEntry, String username)
     {
-        String baseQueryString = String.format("CREATE ROLE IF NOT EXISTS %s WITH LOGIN = %s AND SUPERUSER = %s AND PASSWORD", role, login, superUser);
-        return isPasswordObfuscated ? baseQueryString + PasswordObfuscator.OBFUSCATION_TOKEN : baseQueryString + String.format(" = '%s'", TEST_PW);
+        return String.format("CREATE ROLE IF NOT EXISTS %s WITH PASSWORD = '%s' AND LOGIN = %s AND SUPERUSER = %s",
```
suggestion: keep the `baseQueryString` variable

bereng replied:
I think it reads nicer as it is now. Again personal taste I guess.

ekaterinadimitrova2 replied:
I think we almost always bring the query separately and reads better but I will leave it to you

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/audit/AuditLoggerAuthTest.java:
```diff
@@ -220,6 +220,41 @@ public void testUNAUTHORIZED_ATTEMPTAuditing()
+        logEntry = getInMemAuditLogger().poll();
+        assertLogEntry(logEntry,
+                       AuditLogEntryType.REQUEST_FAILURE,
+                       "ALTER USER " + user
+                       + " WITH PASSWORD "
+                       + PasswordObfuscator.OBFUSCATION_TOKEN
```
I would keep these on one line, similar to the previous  `assertLogEntry`

bereng replied:
+1 auto formatting playing tricks on me.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -135,7 +137,13 @@ else if (e instanceof AuthenticationException)
         }
 
-        builder.appendToOperation(QueryEvents.instance.getObfuscator().obfuscate(e.getMessage()));
+        // A syntax error may reveal the password in the form of 'line 1:33 mismatched input 'secret_password''
+        String safeMsg = e instanceof SyntaxException
+                         && e.getMessage().toLowerCase().contains(PasswordObfuscator.PASSWORD_TOKEN.toLowerCase())
+                                                                                                                   ? "Syntax Exception. Obscured for security reasons."
+                                                                                                                   : e.getMessage();
+
+        builder.appendToOperation(PasswordObfuscator.obfuscate(safeMsg));
```
Please fix the indentation
Also, suggestion (my indentation is also off as GH makes a mess... :-) ):
```
String safeMsg = e instanceof SyntaxException
                         && e.getMessage().toLowerCase().contains(PasswordObfuscator.PASSWORD_TOKEN.toLowerCase())
                         ? "Syntax Exception. Obscured for security reasons."
                         : PasswordObfuscator.obfuscate((e.getMessage()));

        builder.appendToOperation(safeMsg);
```

bereng replied:
Indentation is as per style rules iiuc. But it sure reads weird. I tried an aux variable but being a possible hot path I discarded that. I have re-aligned manually see if that improves things.

ekaterinadimitrova2 replied:
Currently you try to obfuscate -  `Syntax Exception. Obscured for security reasons.` and `e.getMessage()`, both after `Syntax Exception. Obscured for security reasons.` is actually an obfuscation already. What did I miss?

bereng replied:
Maybe I misunderstood you but I am not re-obfuscating. It is a ternary operator that applies only _when_ there is a parsing error to obfuscate the message _in_ the syntax error exception. Apologies if I am being captain obvious or missing the point!

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/cql3/statements/AuthenticationStatement.java:
```diff
@@ -63,5 +65,15 @@ public void checkPermission(ClientState state, Permission required, RoleResource
         }
     }
+
+    public interface CanObfuscatePassword
```
I am not convinced we need an interface, I think I might be missing something in your line of thinking. Can you advise, please? 
Also, I am not sure about the name as it seems just as a regular obfuscater but at the same time we already have `PasswordObfuscater`.

ekaterinadimitrova2 replied:
How about something like `AuthStatementPasswordObfuscator` or something similar?

bereng replied:
I am using the interface bc [here](https://github.com/apache/cassandra/pull/1293/files/c3c9a24a5cf9693ebf4aed23486c6fde5b015a13#diff-df97ca69d481fde559e155c724ca60967a1e57222ea845d8ee8299d7b014df46R252) we need to detect if we can indeed obfuscate only the password. But I am open to suggestions either on implementation or renaming (I am very bad at names lol).

Does it make sense now? should I change the name?

ekaterinadimitrova2 replied:
I need to think about it more. But at least I would probably change the name. We use it also to check statement instance which when we read it as "CanObfuscatePassword" sounds a bit confusing I feel.

bereng replied:
Happy to change it for better wording when we have sthg.

blerer replied:
I would simply add a noop `obfuscatePassword(String query)` method to `AuthenticationStatement` and override it in `CreateRoleStatement` and `AlterRoleStatement`.

bereng replied:
Changed.

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/audit/AuditLoggerAuthTest.java:
```diff
@@ -220,6 +220,41 @@ public void testUNAUTHORIZED_ATTEMPTAuditing()
     }
 
+    @Test
+    public void testBackwardsCompatibleUserCommandObfuscation()
```
Suggestion:
I think it will be better if we just make a few more tests similar to the others (like `testcqlCreateUserAuditing`) and just leave a short comment about backward compatibility if we see a need, WDYT?

bereng replied:
Sorry I am not following. Are you suggesting I rename it? or that I rework them and split them in 3? or both?

I did it all in a single method bc these tests are 'cumulative' sort to speak. If you split them you end up repeating the same operations over and over again. It's less cosmetic but worth it to reduce junit load on the server. I don't have a strong opinion. Whatever you choose is fine by me.

ekaterinadimitrova2 replied:
Well, all others are split and only this one is not. To me the name was confusing as my first thought was - what backward compatibility in audit logging, this is new feature. Then I read and it starts coming but I feel it is not self explanatory.

bereng replied:
This is bc the 'CREATE USER' only works to be backward compatible. 'CREATE ROLE' is the new thing and what should be used instead iiuc. Hence the name. But again tell me what you think is best I don't have a string opinion here.

ekaterinadimitrova2 replied:
I understand the backward compatibility but I wanted to be consistent with other tests and I think it is easier if the test name is more clear. Currently if you read it without the implementation of the test it is not clear what is it actually about.

bereng replied:
Can I resolve this one or is anything pending here?

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/cql3/PasswordObfuscator.java:
```diff
@@ -18,15 +18,20 @@
+    public static final String PASSWORD_TOKEN = PasswordAuthenticator.PASSWORD_KEY.toLowerCase();
 
-    public String obfuscate(String sourceString)
+    public static String obfuscate(String sourceString)
```
Not sure why that was needed.

bereng replied:
Being non-static just added a few locs I saw no reason to have around, the trouble of dragging an instance around and mine came out static so I 'unified' them :shrug:

bereng replied:
I'll take it this is ok to resolve? Feel free to reopen if not. Just tidying things up.

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/audit/AuditLoggerAuthTest.java:
```diff
@@ -150,7 +150,7 @@ public void testCqlALTERRoleAuditing()
         assertTrue(getInMemAuditLogger().size() > 0);
         AuditLogEntry logEntry = getInMemAuditLogger().poll();
-        assertLogEntry(logEntry, AuditLogEntryType.ALTER_ROLE, "ALTER ROLE " + TEST_ROLE + " WITH PASSWORD" + PasswordObfuscator.OBFUSCATION_TOKEN, CASS_USER);
+        assertLogEntry(logEntry, AuditLogEntryType.ALTER_ROLE, "ALTER ROLE " + TEST_ROLE + " WITH PASSWORD = '" + PasswordObfuscator.OBFUSCATION_TOKEN + "'", CASS_USER);
```
I was thinking... let's add a check that an audit log entry does not contain the password further to validating the CQL statement, it will make it more robust in my opinion and it can help preventing some regressions in our entries out of the DCL statement itself.

bereng replied:
Gosh I am stupid, sorry I forgot to mention I had added that in my last push.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/cql3/PasswordObfuscator.java:
```diff
@@ -18,15 +18,20 @@
 /**
  * Obfuscates passwords in a given string
  */
 public class PasswordObfuscator
```
By the way it seems we have a typo. It is "obfuscatEr", I am always confused whether we need 'e' or 'o' in this word :-(

bereng replied:
Oh, I am not native English but googling around I can find `tOr`but not `tEr`. Isn't `tOr` correct?

ekaterinadimitrova2 replied:
I found both and weird enough the spell checker was giving it "er". So...Merriam Webstrer - https://www.merriam-webster.com/dictionary/obfuscator

Sorry for the noise

bereng replied:
Ah no noise at all, always happy to learn.

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/cql3/PasswordObfuscatorTest.java:
```diff
@@ -18,145 +18,223 @@
 {
-    private static final PasswordObfuscator obfuscator = new PasswordObfuscator();
+    private static RoleOptions opts = new RoleOptions();
+    private static String optsPassword = "testpassword";
```
This can be final I think

bereng replied:
Right you are!

blerer commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -135,7 +136,13 @@ else if (e instanceof AuthenticationException)
         }
 
-        builder.appendToOperation(QueryEvents.instance.getObfuscator().obfuscate(e.getMessage()));
+        // A syntax error may reveal the password in the form of 'line 1:33 mismatched input 'secret_password''
+        String safeMsg = e instanceof SyntaxException
+                         && e.getMessage().toLowerCase().contains(PasswordObfuscator.PASSWORD_TOKEN.toLowerCase())
+                             ? "Syntax Exception. Obscured for security reasons."
+                             : e.getMessage();
+
+        builder.appendToOperation(PasswordObfuscator.obfuscate(safeMsg));
```
I do not believe that this logic is really bulletproof.
 
It is not guaranty that the `PASSWORD_TOKEN`  will be part of the message depending on how the original query was formatted. The `ErrorCollector` slicing logic might let that part our depending on where is the syntax issue and how is the query formatted (multiple lines for example).

Another use case that might go undetected by the logic is a typo on the password keyword (eg. pssword)

bereng replied:
Yep that is correct. I thought Levenshtein distance etc but we're on the hot path. Once what you have is just a string this is better than nothing but not bullet proof at all. The challenge is how to improve on that bc I think I can always think of a string that fails to parse that will defeat any logic behind it.

bereng replied:
I tried to improve this in my latest push by checking the query and not the sliced error message. This still fails if:
- There is a typo in the word password
- It is not a DCL stmnt but contains the work password

I am open for suggestions on how to improve this further :thinking:

blerer commented on src/java/org/apache/cassandra/cql3/PasswordObfuscator.java:
```diff
@@ -35,6 +46,25 @@ public String obfuscate(String sourceString)
+        if (!pass.isPresent() || pass.get().isEmpty())
+            return query;
+        else
+            return query.replaceAll(pass.get(), PasswordObfuscator.OBFUSCATION_TOKEN);
```
Using `replaceAll` does not look safe to me. If the password is `role` for example we will end up with a wrongly obfuscated statement that can reveal the actual password.

ekaterinadimitrova2 replied:
Valid point and also if we commit this, it will be a regression to the current way of logging category DCL.

bereng replied:
This is a corner case. We can look if there is a way to avoid it. But I disagree it is a regression. The problem in GA atm is that obscuring queries it's a security problem imo from what I remember from OWASP. It disables the possibility of forensic analysis of logs upon a breach. Attackers use it to hide their steps.

So I see it as an improvement bc only non-parsing queries will suffer that.

blerer replied:
My point was simply that we can do something better than a `replaceAll`. We can find the precise location of the password based on the  `PASSWORD_TOKEN` and simply replace it.

bereng replied:
Aha good point. I will look into that thx.

bereng replied:
In the latest push I have attached an improved regex version of replaceAll that should do the trick imo as `PASSWORD` is a reserved word. if you take into account spaces, newlines, `=` present or not, quotes etc doing it without a regexp seems very complicated. wdyt?

ekaterinadimitrova2 replied:
Tested this one with a few corner cases:

`Type: audit
LogMessage: user:cassandra|host:localhost/127.0.0.1:7000|source:/127.0.0.1|port:60355|timestamp:1636402050002|type:CREATE_ROLE|category:DCL|operation:create role 'superuser' with password = '*******' and superuser = true and login = true;`
We can probably add it as a unit test to cover it and be as a reminder?

bereng replied:
^what is the original cql you wanted added as a junit as I can only see the obscured version please?

bereng commented:
These are all correct observations @blerer But they are an improvement to what we have now. The challenge is improving on what's on this PR or going for the 'correct' solution, which seems to me should be sthg about a one way hash at client side or similar.

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/audit/AuditLoggerAuthTest.java:
```diff
@@ -216,10 +217,47 @@ public void testUNAUTHORIZED_ATTEMPTAuditing()
     }
 
+    @Test
+    public void testBackwardsCompatibleUserCommandObfuscation()
```
```suggestion
   public void testCqlUSERCommandsAuditing()
       //CREATE USER and ALTER USER are supported only for backwards compatibility.
```

bereng replied:
Done except for the space removal which looks like a typo?

blerer commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -135,7 +143,21 @@ else if (e instanceof AuthenticationException)
-        builder.appendToOperation(QueryEvents.instance.getObfuscator().obfuscate(e.getMessage()));
+        // A syntax error may reveal the password in the form of 'line 1:33 mismatched input 'secret_password''
+        String safeMsg = e.getMessage();
+        if (e != null && e instanceof SyntaxException && queries != null && !queries.isEmpty())
```
The `e != null` check can be removed apprently as the line above would have fired a NPE otherwise.

blerer commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -135,7 +143,21 @@ else if (e instanceof AuthenticationException)
+            }
+        }
+
+        builder.appendToOperation(PasswordObfuscator.obfuscate(safeMsg));
```
I would have use extracted the logic creating the message into:
```
public String obfuscatePasswordInformation(Exception e, List<String> queries)
{
    if (e instanceof SyntaxException && queries != null && !queries.isEmpty())
    {
        for (String query : queries)
        {
            if (query.toLowerCase().contains(PasswordObfuscator.PASSWORD_TOKEN))
            {
                return "Syntax Exception. Obscured for security reasons.";
            }
        }
    }

    return PasswordObfuscator.obfuscate(e.getMessage());
}
```

bereng replied:
I'll take that one as optional if you don't mind as I prefer it inline. Total personal thing of mine lol

blerer commented on src/java/org/apache/cassandra/cql3/PasswordObfuscator.java:
```diff
@@ -35,6 +46,26 @@ public String obfuscate(String sourceString)
+        Optional<String> pass = opts.getPassword();
+        if (!pass.isPresent() || pass.get().isEmpty())
+            return query;
+        else
```
No real need for the `else`

blerer commented on src/java/org/apache/cassandra/cql3/QueryEvents.java:
```diff
@@ -250,10 +242,19 @@ public void notifyPrepareFailure(@Nullable CQLStatement statement, String query,
+        String res;
+
+        if (statement == null)
+            res = PasswordObfuscator.obfuscate(query);
```
we can simply return the result it will avoid the need for the else-if

blerer reviewed (commented):
I let a couple of nits otherwise the patch looks good.

bereng committed:
CIRCLE TO MID DO NOT MERGE

bereng commented:
CI [here](https://app.circleci.com/pipelines/github/bereng/cassandra?branch=CASSANDRA-16801-4.0&filter=all)

bereng commented:
CI [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/543/workflows/74532abe-6cb7-4035-aa7a-6b5cbb7fbb10)
CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/543/workflows/0e5d7f6c-7955-4424-a77c-ebcca2c7db88)

Failures are unrelated or we can see they are GH timeouts which worked ok in the 'sister' jobs and in previous runs.

bereng commented:
CI [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/548/workflows/9c848f0a-83d5-4f25-b545-25cf88c36a01)
CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/548/workflows/742b9ee9-65b7-4ba7-aa4a-92259e7950f6)

LGTM

ekaterinadimitrova2 commented on NEWS.txt:
```diff
@@ -37,6 +37,10 @@ using the provided 'sstableupgrade' tool.
 
+New features
+------------
+    - DCL statements in audit logs will now obscure only the password if they don't fail to parse.
```
Is this really a new feature? I mean it sounds weird to me - also, we already were failing to obfuscate in certain parse errors

bereng replied:
Hi @ekaterinadimitrova2 this suggestions from adding this to NEWS.txt comes from the discussion in which you convinced me about adding it :-): `Also NEWS.txt because there will be a visible change for the users. The entries will obfuscate only the password and the DCL statements won't be that useless anymore`. See the second resolved comment from the top.

So I am +1 on it, iiuc Benjamin reviewed as so he is ok with that as well and as you don't seem violently against it I will guess it is ok to leave it. I hope I am reading this ok.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -328,4 +336,21 @@ public void authFailure(QueryState state, Exception cause)
+                if (query.toLowerCase().contains(PasswordObfuscator.PASSWORD_TOKEN))
+                {
+                    return "Syntax Exception. Obscured for security reasons.";
+                }
```
nit: the braces can be removed, can be done on commit

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/cql3/PasswordObfuscator.java:
```diff
@@ -35,6 +46,26 @@ public String obfuscate(String sourceString)
+    }
+
+    /**
+     * Obfuscates the passwords in a query
+     * 
+     * @param query The query whose passwords to obfuscate
+     * @param opts The options containing the password to obfuscate
+     * @return The query with obfuscated passwords
```
```suggestion
     * Obfuscates the password in a query
     * 
     * @param query The query whose password to obfuscate
     * @param opts The options containing the password to obfuscate
     * @return The query with obfuscated password
```

bereng replied:
sgtm

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -328,4 +336,21 @@ public void authFailure(QueryState state, Exception cause)
+            for (String query : queries)
+            {
+                if (query.toLowerCase().contains(PasswordObfuscator.PASSWORD_TOKEN))
+                {
+                    return "Syntax Exception. Obscured for security reasons.";
+                }
```
```suggestion
                    return "Syntax Exception. Obscured for security reasons.";
```

bereng replied:
+1

ekaterinadimitrova2 commented on NEWS.txt:
```diff
@@ -48,7 +52,8 @@ Upgrading
       currently both old and new names should work. Cassandra 4.0.0 and Cassandra 4.0.1 work ONLY with the new names
       (They weren't updated in cassandra.yaml though).
-      
+
+
```
```suggestion

```

bereng replied:
+1

bereng commented:
Rebased, squashed, latest nits added:
CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/570/workflows/3e508eb1-5748-402e-a2ae-5e839623fa85)
CI [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/570/workflows/1cd90fab-07f9-4b69-9eee-b883d44c66ac)

Known failures. LGTM.

bereng commented:
Resolved NEWS.txt conflict
